tests/digest_validation.rs

```diff
@@ -0,0 +1,279 @@
+
+    let mut headers = HeaderMap::new();
+    headers.insert(DIGEST_HEADER, resp_digest.parse().unwrap());
+    headers.insert(
+        "Content-Type",
+        "application/vnd.docker.distribution.manifest.v2+json"
+            .parse()
+            .unwrap(),
+    );
+
+    (headers, MANIFEST)
+}
+
+async fn blob_handler(
+    State(state): State<ServerConfig>,
+    Path(digest): Path<String>,
+) -> Result<(HeaderMap, &'static [u8]), StatusCode> {
+    let (content, resp_digest) = match digest.as_str() {
+        d if d == CONFIG_DIGEST.as_str() => (
+            CONFIG,
+            if state.bad_config {
+                "sha256:deadbeef"
+            } else {
+                CONFIG_DIGEST.as_str()
+            },
+        ),
+        d if d == BLOB_DIGEST.as_str() => (
+            BLOB,
+            if state.bad_blob {
+                "sha256:deadbeef"
+            } else {
+                BLOB_DIGEST.as_str()
+            },
+        ),
+        _ => return Err(StatusCode::NOT_FOUND),
+    };
+
+    let mut headers = HeaderMap::new();
+    headers.insert(DIGEST_HEADER, resp_digest.parse().unwrap());
+
+    Ok((headers, content))
+}
+
+#[derive(Clone, Copy)]
+struct ServerConfig {
+    bad_manifest: bool,
+    bad_config: bool,
+    bad_blob: bool,
+}
+
+struct BadServer {
+    handle: JoinHandle<()>,
+    pub server: String,
+}
+
+impl Drop for BadServer {
+    fn drop(&mut self) {
+        self.handle.abort()
+    }
+}
+
+impl BadServer {
+    pub async fn new(config: ServerConfig) -> Self {
+        let app = Router::new()
+            .route("/v2/busybox/manifests/:digest", get(manifest_handler))
+            .route("/v2/busybox/blobs/:digest", get(blob_handler))
+            .with_state(config);
+
+        let addr = SocketAddr::from(([127, 0, 0, 1], 0));
+        let listener = TcpListener::bind(addr).await.unwrap();
+        let server_addr = listener.local_addr().unwrap();
+        let port = server_addr.port();
+        let server = format!("127.0.0.1:{}", port);
+        let handle = tokio::spawn(async move {
+            axum::serve(listener, app).await.unwrap();
+        });
+        Self { handle, server }
+    }
+}
+
+#[tokio::test]
+async fn test_bad_manifest() {
+    let server = BadServer::new(ServerConfig {
+        bad_manifest: true,
+        bad_config: false,
+        bad_blob: false,
+    })
+    .await;
+
+    let client = Client::new(ClientConfig {
+        protocol: ClientProtocol::Http,
+        platform_resolver: Some(Box::new(linux_amd64_resolver)),
+        ..Default::default()
+    });
+    let auth = &oci_distribution::secrets::RegistryAuth::Anonymous;
+
+    let reference = Reference::try_from(format!(
+        "{}/busybox@sha256:deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef",
+        server.server
+    ))
+    .expect("failed to parse reference");
+
+    client
+        .pull_manifest(&reference, auth)
+        .await
+        .expect_err("Expected an error with a mismatched sha");
```

Comment:
With typed errors we could write an assertion about the specific error type we get back